File: interviews/algorithm/java17/src/main/java/hoa/can/code/ListNode.java
package hoa.can.code;

import java.util.Objects;

public class ListNode {
    // shut up let me focus on reverting linked list
    public int val;
    public ListNode next;

    ListNode() {
    }

    public ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        }

        if (!(obj instanceof ListNode)) {
            return false;
        }

        ListNode other = (ListNode) obj;
        return this.val == other.val && Objects.equals(this.next, other.next);
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.val, this.next);
    }
}

File: interview_prep/algorithm/Kotlin/src/main/kotlin/hoa/can/code/ListNode.kt
package hoa.can.code

//data class ListNode (var value: Int, var next: ListNode)
class ListNode(var value: Int) {
    var next: ListNode? = null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ListNode) return false

        return value == other.value && next == other.next
    }

    override fun hashCode(): Int {
        var result = value
        result = 31 * result + (next?.hashCode() ?: 0)
        return result
    }
}
